admin/services/data-access/pupil-census-import.data.service.js

```diff
@@ -0,0 +1,69 @@
+'use strict'
+
+const { TYPES } = require('tedious')
+const moment = require('moment')
+
+const config = require('../../config')
+const pupilCensusImportDataService = {}
+const sqlService = require('./sql.service')
+const sqlPoolService = require('./sql.pool.service')
+
+/**
+ * Execute pupil data bulk import
+ * @param {Array} pupilData
+ * @param {Array} schools
+ * @return {Promise<*>}
+ */
+pupilCensusImportDataService.sqlBulkImport = async(pupilData, schools) => {
+  const result = {}
+  const con = await sqlPoolService.getConnection()
+
+  try {
+    result.output = await bulkLoadData(con, pupilData, schools)
+  } catch (error) {
+    result.errorOutput = error
+  }
+  con.release()
+  return result
+}
+
+const bulkLoadData = (connection, pupilData, schools) => {
+  return new Promise((resolve, reject) => {
+    const bulkLoad = connection.newBulkLoad(`${config.Sql.Database}.${sqlService.adminSchema}.[pupil]`, function (error, rowCount) {
+      if (error) {
+        reject(error)
+      }
+      resolve(`Inserted ${rowCount} rows`)
+    })
+
+    bulkLoad.addColumn('school_id', TYPES.Int, {nullable: false})
+    bulkLoad.addColumn('upn', TYPES.Char, {length: 13, nullable: false})
+    bulkLoad.addColumn('lastName', TYPES.NVarChar, {length: 'max', nullable: false})
+    bulkLoad.addColumn('foreName', TYPES.NVarChar, {length: 'max', nullable: false})
+    bulkLoad.addColumn('middleNames', TYPES.NVarChar, {length: 'max', nullable: true})
+    bulkLoad.addColumn('gender', TYPES.Char, {length: 1, nullable: false})
+    bulkLoad.addColumn('dateOfBirth', TYPES.DateTimeOffset, {nullable: false})
+
+    for (let index = 0; index < pupilData.length; index++) {
+      const csvRow = pupilData[index]
+      const dfeNumber = `${csvRow[0]}${csvRow[1]}`
+      const school = schools.find(s => s.dfeNumber === parseInt(dfeNumber))
```

Comment:
Should be a hash lookup, not a find.

Reply:
@jon-shipley Updated since performance should be taken greatly into account for this lookup